ktor-utils/common/src/io/ktor/util/Attributes.kt

```diff
@@ -4,24 +4,26 @@
 
 package io.ktor.util
 
+import kotlin.jvm.*
 import kotlin.reflect.*
 
 /**
  * Specifies a key for an attribute in [Attributes]
  * @param T is a type of the value stored in the attribute
  * @param name is a name of the attribute for diagnostic purposes. Can't be blank
  */
+@JvmSynthetic
 public inline fun <reified T : Any> AttributeKey(name: String): AttributeKey<T> =
-    AttributeKey(name, T::class.toString())
+    AttributeKey(name, typeOf<T>())
 
 /**
  * Specifies a key for an attribute in [Attributes]
  * @param T is a type of the value stored in the attribute
  * @param name is a name of the attribute for diagnostic purposes. Can't be blank
  */
-public class AttributeKey<T : Any> @PublishedApi internal constructor(
+public data class AttributeKey<T : Any> @JvmOverloads constructor(
     public val name: String,
-    private val type: String
+    private val type: KType = typeOf<Any>(),
```

Comment:
I updated the class to keep the type information here.  I think it could be useful for the dependency injection story, and maybe for better forensics when the types don't align.

The constructor also needed to be exposed for when the KType can't be inferred, specifically for the example above with ConditionalHeaders (the suspend keyword breaks the type inference).